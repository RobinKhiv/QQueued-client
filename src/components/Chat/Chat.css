.chatRoomContainer {
  display: flex;
  justify-content: center;
}

.chatRoomInput {
  position: relative;
  text-align: center;
  border: 2px solid black;
  width: 50%;
  margin: 20px;
  border-radius: 25px;
  background: white;
}

.userHasJoined {
  background: #5e0a0a;
  color: white;
}

.chatRoomForm,
.chatMessage {
  background: none;
}

.chatMessage {
  display: flex;
  flex-direction: row-reverse;
}

.chatRoomButton {
  background: white;
}

.currentMessage {
  background: radial-gradient(5px -9px, circle, white 4%, #5ca0d3 26px);
  padding: 5px;
  border: 1px solid black;
  border-radius: 12px; /* one half of ( (border * 2) + height + padding ) */
  box-shadow: 2px 2px 2px darkgray;
  color: white;
  font: bold 15px/13px Helvetica, Verdana, Tahoma;
  text-align: left;
  width: 40%;
  line-height: 20px;
}

.sendMessage {
  width: 75%;
  padding: 10px;
  border-radius: 25px;
  background: #c8e6f5;
}

.currentUser {
  width: 30px;
  height: 30px;
  border-radius: 50%;
  color: white;
  font-weight: bolder;
  font-size: 15px;
  line-height: 200%;
  text-align: center;
  background: #5e0a0a;
  margin: 5px;
}
