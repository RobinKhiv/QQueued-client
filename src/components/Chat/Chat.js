import React, { Component } from "react";
import openSocket from "socket.io-client";
import "./Chat.css";
import Button from "../Button/Button";
import QueueContext from "../../context/QueueContext";
import apiSerivce from "../../services/api-service";

export default class Chat extends Component {
  static contextType = QueueContext;
  constructor(props) {
    super(props);

    this.state = {
      input: "",
      users: [],
      messages: [],
      to: [],
      studentInput: "",
      student: ""
    };
    this.socket = openSocket(
      /* config.API_ENDPOINT ||  */ "http://localhost:8000"
    );
    this.rooms = [];
    this.recevier = "";
  }

  scrollToBottom() {
    this.messagesEnd.scrollIntoView({ behavior: "smooth" });
  }

  componentDidUpdate() {
    this.scrollToBottom();
  }

  async componentDidMount() {
    if(this.focusInput){
      this.focusInput.focus()
    }
    const { currentlyBeingHelped } = await apiSerivce.getQueue();
    const filtedList = currentlyBeingHelped.filter(
      i =>
        i.studentName === this.props.user.user.full_name ||
        i.mentorName === this.props.user.user.full_name
    );

    this.socket.emit("join-room", {
      userName: this.props.user.user.full_name,
      list: filtedList[filtedList.length - 1]
    });
    this.socket.on("entered", data => {
      this.rooms.push(data);
      this.setState({ users: [...this.state.users, data.userName] });
    });

    this.socket.on("message", data => {
      this.setState({
        messages: [...this.state.messages, data],
        student: "",
        studentInput: ""
      });
    });

    this.socket.on("isTyping", data => {
      if (this.state.users[0] !== data.to.studentName) {
        this.setState({
          student: data.to.studentName,
          studentInput: data.type
        });
      }
    });
    this.scrollToBottom();
  }
  componentWillUnmount() {
    const helpedArray = this.context.currentlyBeingHelped;
    this.socket.emit("left", {
      user: this.state.users[0],
      room: this.rooms,
      to: helpedArray[helpedArray.length - 1]
    });
    this.socket.close();
  }

  handleChange = e => {
    const helpedArray = this.context.currentlyBeingHelped.filter(
      i =>
        i.studentName === this.state.users[0] ||
        i.mentorName === this.state.users[0]
    );
    this.socket.emit("isTyping", {
      type: e,
      to: helpedArray[helpedArray.length - 1],
      user: this.state.users[0]
    });
    this.setState({ input: e });
  };

  handleSubmit = e => {
    e.preventDefault();
    const helpedArray = this.context.currentlyBeingHelped.filter(
      i =>
        i.studentName === this.state.users[0] ||
        i.mentorName === this.state.users[0]
    );
    this.socket.emit("message", {
      user: this.state.users[0],
      text: this.state.input,
      time: new Date().toLocaleTimeString(),
      room: this.rooms,
      to: helpedArray[helpedArray.length - 1]
    });
    this.setState({ student: "", studentInput: "" });
    e.target.reset();
  };

  render() {
    let thread;
    if (this.state.messages.length > 0) {
      thread = this.state.messages.map((i, j) => {
        if (this.props.user.user.full_name !== i.user) {
          return (
            <div className="foreignChatMessage" key={j}>
              <span title={i.user} className="foreignUser">
                {i.user.charAt(0).toUpperCase()}
              </span>
              <p className="foreignMessage">{i.text}</p>
              <p className="time">{i.time}</p>
            </div>
          );
        }

        return (
          <div key={j} className="chatMessage">
            <span title={i.user} className="currentUser">
              {i.user.charAt(0).toUpperCase()}
            </span>
            <p className="currentMessage">{i.text}</p>
            <p className="time">{i.time}</p>
          </div>
        );
      });
    }
    let activeUsers;
    if (this.state.users.length > 0) {
      activeUsers = this.state.users.map((i, j) => {
        return (
          <p className="userHasJoined" key={j}>
            {i} has joined
          </p>
        );
      });
    }
    return (
      <div className="chatRoomContainer">
        <section className="chatRoomInput">
          <section className="messages">
            {activeUsers}
            {thread}
            {this.state.studentInput && (
              <div>
                <span>{this.state.student} preview: </span>
               <p className="foreignMessage">{this.state.studentInput}</p> 
              </div>
            )}
            <div
              style={{ float: "left", clear: "both" }}
              ref={el => {
                this.messagesEnd = el;
              }}
            />
          </section>
          <form className="chatRoomForm" onSubmit={e => this.handleSubmit(e)}>
            <input
              ref={(input) => {this.focusInput = input}}
              className="sendMessage"
              type="text"
              name="input-field"
              id=""
              placeholder="send a message"
              onChange={e => this.handleChange(e.target.value)}
              autoComplete="off"
            />
            <Button type="submit">Send</Button>
          </form>
        </section>
      </div>
    );
  }
}
