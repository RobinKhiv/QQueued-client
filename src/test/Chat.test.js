import React from "react";
import '../setupTests'
import { MemoryRouter } from 'react-router-dom';
import ReactDOM from "react-dom";
import Chat from '../components/Chat/Chat';
import TestRenderer from 'react-test-renderer';
import { mount } from 'enzyme';


describe('Chat class component', () => {
<<<<<<< HEAD
  it('should render Input', () => {
    const wrapper = mount(<Chat />);
    expect(wrapper.find('.sendMessage')).toHaveLength(1)
  })

  it('should render Input', () => {
    const wrapper = mount(<Chat />);
    const mockRef = {}
    expect(wrapper.ref(mockRef))
  })

  it('renders the UI as expected', () => {
    const tree = TestRenderer.create(<MemoryRouter><Chat /></MemoryRouter>).toJSON()
    expect(tree).toMatchSnapshot()
=======
  it('renders without crashing', () => {
    const testChat = TestRenderer.create(
      <MemoryRouter>
        <Chat />
      </MemoryRouter>
    );
    
    expect(testChat.getInstance())
    testChat.unmount()
>>>>>>> 849a9b91ebb189095716b83b150a1148e7927a18
  })
})