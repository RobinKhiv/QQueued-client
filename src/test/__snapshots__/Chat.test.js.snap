// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chat class component renders the UI as expected 1`] = `
<div
  className="chatRoomContainer"
>
  <section
    className="chatRoomInput"
  >
    <section
      className="messages"
    >
      
      <div
        style={
          Object {
            "clear": "both",
            "float": "left",
          }
        }
      />
    </section>
    <form
      className="chatRoomForm"
      onSubmit={[Function]}
    >
      <input
        autoComplete="off"
        className="sendMessage"
        id=""
        name="input-field"
        onChange={[Function]}
        placeholder="send a message"
        type="text"
      />
      <button
        className="Button"
        type="submit"
      >
        Send
      </button>
    </form>
  </section>
</div>
`;
